### PR TITLE
Local mode v2

### PR DESCRIPTION
**Pull request by CameronDiver**
This PR adds support for running multicontainer local mode applications.

### CODE REVIEW

File: lib/utils/device/api.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+const deviceEndpoints = {
+	setTargetState: 'v2/local/target-state',
+	getTargetState: 'v2/local/target-state',
+	getDeviceInformation: 'v2/local/device-info',
+	logs: 'v2/local/logs',
+	ping: 'ping',
+	version: 'v2/version',
+};
+
+export class DeviceAPI {
```
[the remark this answers is not in the document]
Comment (CameronDiver) [reply lib/utils/device/api.ts]: This is a great idea! Very keen for this.

File: lib/utils/device/deploy.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+	const targetState = generateTargetState(
+		currentTargetState,
+		project.composition,
+	);
+	logger.logDebug(`Sending target state: ${JSON.stringify(targetState)}`);
+
+	await api.setTargetState(targetState);
+
+	console.log();
```
[the remark this answers is not in the document]
Comment (CameronDiver) [reply lib/utils/device/deploy.ts]: So this is to add a newline between the build output and the log output. Will add a comment.

File: lib/actions/push.ts [hunk 3]
```diff
@@ -75,27 +101,36 @@ async function getAppOwner(sdk: ResinSDK, appName: string) {
...
+
 		Examples:
 
 			$ resin push myApp
 			$ resin push myApp --source <source directory>
 			$ resin push myApp -s <source directory>
+
+			$ resin push 10.0.0.1
+			$ resin push 10.0.0.1 --source <source directory>
+			$ resin push 10.0.0.1 -s <source directory>
```
[the remark this answers is not in the document]
Comment (pimterry) [reply lib/actions/push.ts]: (Discussed, and we're going to stick with this format)

File: lib/actions/push.ts [hunk 4]
```diff
@@ -15,8 +15,34 @@ limitations under the License.
...
+	Device,
+}
+
+function getBuildTarget(appOrDevice: string): BuildTarget | null {
+	// First try the application regex from the api
+	if (/^[a-zA-Z0-9_-]+$/.test(appOrDevice)) {
+		return BuildTarget.Cloud;
+	}
+
+	// Now try to detect an IP address, regex from:
```
Comment (pimterry) [inline lib/actions/push.ts]: Nit: This comment has gone funky

File: lib/actions/push.ts [hunk 5]
```diff
@@ -75,27 +101,38 @@ async function getAppOwner(sdk: ResinSDK, appName: string) {
...
 		This command can be used to start a build on the remote
-		resin.io cloud builders. The given source directory will be sent to the
+		resin.io cloud builders, or a local mode resin device.
+
+		When building on the resin cloud the given source directory will be sent to the
 		resin.io builder, and the build will proceed. This can be used as a drop-in
 		replacement for git push to deploy.
 
+		When building on a local mode device, the given source directory will be built on
+		device, and the resulting contianers will be run on the device. Logs will be
```
Comment (pimterry) [inline lib/actions/push.ts]: Typo: `contianers` -> `containers`

File: lib/utils/device/deploy.ts [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+export interface DeviceDeployOptions {
+	source: string;
+	deviceHost: string;
+	devicePort?: number;
+}
+
+async function checkSource(source: string): Promise<boolean> {
+	const { fs } = await import('mz');
+	return (await fs.exists(source)) && (await fs.stat(source)).isDirectory();
```
Comment (pimterry) [inline lib/utils/device/deploy.ts]: Nit: you could do these in parallel and it would be _sliiiightly_ faster.

**Pull request merged**